### PR TITLE
feat: save provider error response in repository

### PR DESCRIPTION
**Pull request by kshitij-k-osmosys**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](https://github.com/OsmosysSoftware/osmo-x/blob/main/CONTRIBUTING.md#-submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [x] I have performed unit testing for the new feature added or updated to ensure the new features added are working as expected.
- [ ] I have added/updated test cases to the [test suite](https://github.com/OsmosysSoftware/osmo-x/blob/main/apps/api/OsmoX.postman_collection.json) as applicable.
- [ ] I have performed preliminary testing using the [test suite](https://github.com/OsmosysSoftware/osmo-x/blob/main/apps/api/OsmoX.postman_collection.json) to ensure that any existing features are not impacted and any new features are working as expected as a whole.
- [ ] I have added/updated the required [api docs](https://github.com/OsmosysSoftware/osmo-x/tree/main/apps/api/docs) as applicable.
- [ ] I have added/updated the `.env.example` file with the required values as applicable.

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [x] Screenshots have been added (optional)
- [ ] Query request and response examples have been added (as applicable, in case added or updated)
- [ ] Documentation changes have been listed (as applicable)
- [ ] Test suite/unit testing output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

_Note: Reviewer should ensure that the checklist and description have been populated and followed correctly, and the PR should be merged only after resolving all conversations and verifying that CI checks pass._

---

**Description:**

Save error from other the 3rd party service response in the DB

**Related changes:**

- Test 360Dialog response directly from the provider
- Check response we are saving, then update the response we save to include error.response.data whenever available
- Start checking other providers and add a catch block if necessary to capture the response sent by the provider itself
- Check direct responses for Twilio and Plivo
- Compare to the responses we store in repo and make a note of possible keys we can show
- Add more details in the final response object we store in repository
- Add error handling for Plivo

**Screenshots:**
NA

**Query request and response:**
NA

**Documentation changes:**
NA

**Test suite/unit testing output:**
NA

**Pending actions:**
NA

**Additional notes:**
NA

## Summary by CodeRabbit

- **Bug Fixes**
	- Enhanced error handling and retry logic for notification processing, providing more detailed error information.
	- Improved consistency in storing results from notification providers.
	- Improved error handling in SMS sending, allowing for more informative logging.

- **New Features**
	- Updated promise handling in message sending to improve reliability and error management.
	- Enhanced error logging for better traceability of notification responses.

### CODE REVIEW

File: apps/api/src/modules/providers/wa360dialog/wa360dialog.service.ts [hunk 1]
```diff
@@ -61,13 +62,30 @@ export class Wa360dialogService {
...
+          data: error.response.data,
+        };
+        // Log relevant parts of the error response
+        this.logger.error('Error from provider:', providerResponseError);
+
+        throw error;
+      } else {
+        // Handle cases where there is no response (network issues, etc.)
+        throw new Error(`Unexpected error occurred: ${error}`);
+      }
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/providers/wa360dialog/wa360dialog.service.ts]: fixed 3a5d75bb3f03be88c80c1211eecc6703dd9bc867

File: apps/api/src/modules/providers/wa360dialog/wa360dialog.service.ts [hunk 2]
```diff
@@ -61,13 +62,30 @@
...
+      return response.data;
+    } catch (error) {
+      if (error.response) {
+        const providerResponseError = {
+          status: error.response.status,
+          statusText: error.response.statusText,
+          data: error.response.data,
+        };
+        // Log relevant parts of the error response
+        this.logger.error('Error from provider:', providerResponseError);
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/providers/wa360dialog/wa360dialog.service.ts]: fixed: 3a5d75bb3f03be88c80c1211eecc6703dd9bc867

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 3]
```diff
@@ -78,7 +78,14 @@ export class SmsPlivoService {
 
       return response;
     } catch (error) {
-      throw new Error(`Failed to send message: ${error.message}`);
+      if (error.apiID) {
+        // Log relevant parts of the error response
+        this.logger.error(`Error sent from provider: ${providerId}`, error.message);
+        throw error;
```
Comment (xixas) [inline apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts:84]: I am assuming whereever we are throwing errors we have it being handled in parent classes?

File: apps/api/src/jobs/consumers/notifications/notification.consumer.ts [hunk 4]
```diff
@@ -97,12 +97,26 @@ export abstract class NotificationConsumer {
       }
 
       this.logger.debug(`Updating result of notification with id ${notification.id}`);
-      notification.result = { result: { message: error.message, stack: error.stack } };
+      notification.result = {
+        result: {
+          message: error.message,
```
Comment (xixas) [inline apps/api/src/jobs/consumers/notifications/notification.consumer.ts:102]: @kshitij-k-osmosys  Can you screenshot of all these different responses.

File: apps/api/src/jobs/consumers/notifications/notification.consumer.ts [hunk 5]
```diff
@@ -97,12 +97,26 @@ export abstract class NotificationConsumer {
...
+          apiID: error.apiID, // Plivo
+          stack: error.stack,
+        },
+      };
       this.logger.error(`Error sending notification with id: ${id}`);
       this.logger.error(JSON.stringify(error, ['message', 'stack'], 2));
 
       // Save retry attempt record
-      await this.saveRetryAttempt(notification, { message: error.message, stack: error.stack });
+      await this.saveRetryAttempt(notification, {
```
Comment (xixas) [inline apps/api/src/jobs/consumers/notifications/notification.consumer.ts:113]: Why do we have to send result as second option? Don't we already have it in notication.result?

**Review by xixas — CHANGES_REQUESTED** [review]

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts (hunk 3 above)
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts:84]: Could you elaborate a bit more and tell the following answers the query?

Each provider has a send methed, result of which is sent to `processNotificationQueue` present in `src/jobs/consumers/notifications/notification.consumer.ts`
This has a catch block for handling errors

**Review by xixas — APPROVED** [review]

**Pull request merged**